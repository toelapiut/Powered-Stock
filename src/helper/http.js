import axios from 'axios';

export const quandl = axios.create({
  baseURL: process.env.process.env.REACT_APP_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
    'Allow': 'GET, PUT, HEAD, OPTIONS'
  }
});

quandl.interceptors.request.use((config) => {
  config.params = config.params || {};
  config.params['key'] = process.env.REACT_APP_API_KEY;
  return config;
});
