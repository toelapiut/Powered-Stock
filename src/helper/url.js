export const url = {
  markets: 'markets',
  stocks: 'stocks/'
};