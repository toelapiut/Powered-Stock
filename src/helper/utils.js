import humps from 'humps';
import checkPropTypes from 'check-prop-types';
import {timeFormat} from 'd3-time-format';
import {bisector} from 'd3-array';
import moment from 'moment';

export const transpose = (rows, cols) => {
  /*
   * check if the row and col are empty and throw an error to
   * inform of the necessary requirements
   */
  if (typeof rows === 'undefined' || typeof cols === 'undefined') {
    throw new Error('row and col can not be undefined');
  }

  /*
   * throw error if the array isn't an array or col isn't an array
   */
  if (!Array.isArray(rows)) throw new Error('row must be  array');
  if (!Array.isArray(cols)) throw new Error('col must be  array');

  let dataset = [];

  for (let row of rows) {
    /*
     * check the data type before iterating through the matrix
     */
    if (!Array.isArray(row)) throw new Error('row must be a matrix row: [[a, b]]');
    let obj = {};
    row.forEach((item, index) => {
      obj[cols[index]['name']] = item;
    });
    dataset.push(obj);
  }

  return dataset;
};


export const transposeStock = (rows) => {
  let data = [];
  for (let row of rows) {
    let obj = {};
    const [date, close] = row;
    obj['date'] = date;
    obj['close'] = close;
    data.push(obj);
  }
  return data;
};

export const toCamelCase = (item) => {
  if (typeof item === 'undefined') {
    return item;
  }
  return humps.camelizeKeys(item);
};

export const checkPositivity = (num) => {
  if (typeof num === 'undefined') throw new Error('argument must not be empty');
  if (typeof num !== 'number') throw new Error('argument is not a number');
  let check = Math.sign(Number(num));
  return check !== -1;
};

export const formatDate = timeFormat('%b %d, \'%y');

export const getDate = (d) => new Date(d.date);

export const getStockValue = (d) => d.close;

export const bisectDate = bisector(d => new Date(d.date)).left;

export const getWindowDimensions = () => {
  const {innerWidth: width, innerHeight: height} = window;
  return {
    width,
    height
  };
};

export const dateFormat = (date, format = 'DD MMM, YYYY') => {
  return moment(date).format(format);
};

export const findTestByAttr = (component, attr) => {
  return component.find(`[data-test='${attr}']`);
};

export const checkProps = (component, expectedProps) => {
  checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
};

