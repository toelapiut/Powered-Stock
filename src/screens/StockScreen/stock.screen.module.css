.container {
    @apply mx-6
}

.areaClose {
    @apply bg-koamaru p-3 rounded-lg my-6
}