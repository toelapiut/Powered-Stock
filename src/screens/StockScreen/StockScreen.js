import React from 'react';
import PropTypes from 'prop-types';
import AreaClose from '../../components/AreaClose';
import Header from '../../components/Header';
import styles from './stock.screen.module.css';
import {Error} from '../Error/Error';
import {Loading} from '../../components/Loading/Loading';
import {getWindowDimensions} from '../../helper/utils';

let dimensions = getWindowDimensions();
let width = dimensions.width - 450;
let height = dimensions.height - 150;
export const StockScreen = ({name, error, loading, date, onOpenCalendar, isOpen, ticker, start, end, onChangeDates, stocks}) => {

  if (typeof error !== 'undefined') {
    return (
      <div data-test={'error-container'}>
        <Error/>
      </div>
    );
  }

  if (loading) {
    return (
      <div className={styles.loader} style={{height: height, width: width}} data-test={'loading-container'}>
        <Loading/>
      </div>
    );
  }

  return (
    <div className={styles.container} data-test={'stock-screen-container'}>
      <Header
        date={date}
        isOpen={isOpen}
        onChangeDates={onChangeDates}
        name={name}
        ticker={ticker}
        start={start}
        end={end}
        onOpenCalendar={onOpenCalendar}
      />
      <div className={styles.areaClose}>
        {stocks.length > 1 &&
          <div data-test={'area-close-container'}>
            <AreaClose
              stock={stocks}
            />
          </div>
        }
      </div>
    </div>
  );
};

StockScreen.propTypes = {
  name: PropTypes.string.isRequired,
  error: PropTypes.object,
  loading: PropTypes.bool.isRequired,
  ticker: PropTypes.string.isRequired,
  date: PropTypes.shape({
    start: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.object,
    ]),
    end: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.object,
    ])
  }),
  start: PropTypes.string,
  end: PropTypes.string,
  onOpenCalendar: PropTypes.func.isRequired,
  onChangeDates: PropTypes.func.isRequired,
  stocks: PropTypes.array.isRequired,
  isOpen: PropTypes.bool.isRequired,
};