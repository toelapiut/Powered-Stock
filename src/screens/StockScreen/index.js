import {StockScreen} from './StockScreen';

export default StockScreen;