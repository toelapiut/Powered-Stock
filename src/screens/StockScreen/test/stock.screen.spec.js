import React from 'react';
import StockScreen from '../';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import {findTestByAttr, checkProps} from '../../../helper/utils';


const setUp = (props = {}) => {
  return shallow(<StockScreen {...props}/>);
};

let props = {
  error:undefined,
  loading:false,
  date: {
    start: '01-04-2020',
    end: '01-04-2020'
  },
  name: 'Powered By People Inc',
  ticker: 'PBP',
  start: '01-04-2020',
  end: '01-04-2020',
  isOpen: false,
  onChangeDates: () => null,
  onOpenCalendar: () => null,
  stocks: [
    {
      'date': '1980-12-31',
      'close': 34.13
    },
    {
      'date': '1981-01-31',
      'close': 28.25
    },

    {
      'date': '2018-03-31',
      'close': 168.34
    }
  ],
};

describe('StockScreen Component', () => {
  describe('StockScreen renders tests', () => {
    let component;
    beforeAll(() => {
      component = setUp(props);
    });

    test('should render snapshot correctly', () => {
      expect(toJson(component)).toMatchSnapshot();
    });

    it('should render without error', () => {
      const wrapper = findTestByAttr(component, 'stock-screen-container');
      expect(wrapper.length).toBe(1);
    });
  });

  describe('Check property types', () => {
    it('should not throw a warning', () => {
      const propErr = checkProps(StockScreen, props);
      expect(propErr).toBeUndefined();
    });
  });

  describe('renders loading page', () => {
    let component;
    let calendarProps = {
      ...props,
      loading: true
    };

    beforeAll(() => {
      component = setUp(calendarProps);
    });

    it('should render loading component without error', () => {
      const wrapper = findTestByAttr(component, 'loading-container');
      expect(wrapper.length).toBe(1);
    });

    it('should not render stock-screen-container', () => {
      const wrapper = findTestByAttr(component, 'stock-screen-container');
      expect(wrapper.length).toBe(0);
    });
  });

  describe('renders error page', () => {
    let component;
    let calendarProps = {
      ...props,
      error:{},
    };

    beforeAll(() => {
      component = setUp(calendarProps);
    });

    it('should render Error component', () => {
      const wrapper = findTestByAttr(component, 'error-container');
      expect(wrapper.length).toBe(1);
    });
  });

  describe('Areaclose component render', () => {
    let component;
    let calendarProps = {
      ...props,
      stocks: []
    };

    beforeAll(() => {
      component = setUp(calendarProps);
    });

    it('should not render areaclosed component', () => {
      const wrapper = findTestByAttr(component, 'area-close-container');
      expect(wrapper.length).toBe(0);
    });
  });
});