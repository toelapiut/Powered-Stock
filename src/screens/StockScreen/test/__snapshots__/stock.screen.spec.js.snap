// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StockScreen Component StockScreen renders tests should render snapshot correctly 1`] = `
<div
  className="container"
  data-test="stock-screen-container"
>
  <Header
    date={
      Object {
        "end": "01-04-2020",
        "start": "01-04-2020",
      }
    }
    end="01-04-2020"
    isOpen={false}
    name="Powered By People Inc"
    onChangeDates={[Function]}
    onOpenCalendar={[Function]}
    start="01-04-2020"
    ticker="PBP"
  />
  <div
    className="areaClose"
  >
    <div
      data-test="area-close-container"
    >
      <WrappedComponent
        stock={
          Array [
            Object {
              "close": 34.13,
              "date": "1980-12-31",
            },
            Object {
              "close": 28.25,
              "date": "1981-01-31",
            },
            Object {
              "close": 168.34,
              "date": "2018-03-31",
            },
          ]
        }
      />
    </div>
  </div>
</div>
`;
